### PR TITLE
Expose sync manager component to iOS

### PR DESCRIPTION
**Pull request by lougeniaC64**
This PR, in addition to (PR #94)[https://github.com/mozilla/rust-components-swift/pull/94] in `rust-components-swift`, expose the sync manager component to iOS for the sync manager integration. In also adds the following:

- `RustSyncTelemetryPing.Swift` and tests, similar to `SyncTelemetryPing.kt` for deserializing the json telemetry data in `SyncResult`
- Renaming sync manager's `DeviceType` to avoid a naming collision with other a-s components and code to avoid the breaking change for android components

This PR should not be merged until a PR to add namespacing to iOS's `SyncReason`, `SyncResult`, and `SyncManager` types is merged.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[ff-android: firefox-android-branch-name]` and/or `[fenix: fenix-branch-name]` to the PR title.

### CODE REVIEW

File: components/sync_manager/src/syncmanager.udl
Comment (mhammond) [inline components/sync_manager/src/syncmanager.udl:59]: we might be able to kill this - it looks like iOS already doesn't do the right thing here so we prefer the fxa record for the name. IOW, there's probably no need to actually sync when the name changes.

File: components/sync_manager/src/syncmanager.udl
Comment (mhammond) [inline components/sync_manager/src/syncmanager.udl:58]: Android treats `DidLogin` and `EnabledChange` as the same - https://github.com/mozilla-mobile/firefox-android/blob/5a961111addca08dcaa402d2a79b05920c04e731/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/Types.kt#L42. I wonder if we should just do the same here? Otherwise, we should probably get a PR/issue up for Android to use this new one (and given the above link, that seems like it would be a trivial change for them)

File: components/sync15/ios/RustSyncTelemetryPing.swift
Comment (mhammond) [inline components/sync15/ios/RustSyncTelemetryPing.swift:2]: sheesh, it really sucks that we still need this hackery (but yeah, I agree we do - it just sucks :)

**Review by mhammond — COMMENTED** [review]
looking good!

File: components/sync15/ios/RustSyncTelemetryPing.swift
Comment (lougeniaC64) [reply components/sync15/ios/RustSyncTelemetryPing.swift:2]: Yeah, sadly--particularly if we want to create more/better graphs with this telemetry data. I also think I might need to add some unit tests here (although there aren't any for the android logic this file was based on).

**Comment by lougeniaC64** [comment]
> looking good!

Thanks! One thing I'll note is that I am holding off on addressing the breaking changes created for android by the UDL renaming of `DeviceType`, `SyncReason`, and `SyncResult` until the local testing of syncing with the component is finished. Adding those hacks will probably be one of the last things I do before officially opening this PR up for review so that I can reduce the amount of churn here.

File: components/sync_manager/src/syncmanager.udl
Comment (lougeniaC64) [reply components/sync_manager/src/syncmanager.udl:59]: Yeah actually these enum variants are a relic from my first pass at the sync manager work. `ClientNameChanged` is a variant in the native `SyncReason` enum but it's only used by the swift client synchronizer code, which will ultimately be removed. I think I'm going to remove it here and just map it to `EnabledChange` where needed.

File: components/sync_manager/src/syncmanager.udl
Comment (lougeniaC64) [reply components/sync_manager/src/syncmanager.udl:58]: Also `DidLogin` is also mainly used by the old sync code so I think I'm going to remove it to and map `SyncReason.DidLogin` to `RustSyncReason.EnabledChange`.

File: components/sync15/ios/RustSyncTelemetryPing.swift
Comment (tarikeshaq) [inline components/sync15/ios/RustSyncTelemetryPing.swift:7]: Swift might have nice support for deserializable types, what do you think about using something like [`Codable`](https://developer.apple.com/documentation/swift/codable) and [`JSONDecoder`](https://developer.apple.com/documentation/foundation/jsondecoder)?

Comment (tarikeshaq) [reply components/sync15/ios/RustSyncTelemetryPing.swift:7]: The more I look at it the more I see custom logic that `Codable` won't give us by default... We can implement the `init(from decoder: Decoder)` but I'm not sure that's worth it..

File: components/sync15/ios/RustSyncTelemetryPing.swift
Comment (tarikeshaq) [inline components/sync15/ios/RustSyncTelemetryPing.swift:8]: nit should those be `let` and not `var`?

File: components/sync15/ios/RustSyncTelemetryPing.swift
Comment (tarikeshaq) [inline components/sync15/ios/RustSyncTelemetryPing.swift:60]: hmm do we need all the `toJSON` functions? do we expect this type to also be serializable?

File: components/sync_manager/src/syncmanager.udl
Comment (tarikeshaq) [inline components/sync_manager/src/syncmanager.udl:57]: Could this be a breaking change in android? in the case where android depends on an exhaustive `when` expression that checks all the variants(I'm not sure if does or doesn't, just making sure we mark it as such if it is 😬)

File: megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/RustSyncTelemetryPingTests.swift
Comment (tarikeshaq) [inline megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/RustSyncTelemetryPingTests.swift:14]: do we need more tests for this, or do we think it's good enough to test the happy path?

**Review by tarikeshaq — COMMENTED** [review]
This is looking great!! 🚀 


Dropped a couple of thoughts as I was skimming through but looks good!

File: components/sync_manager/src/syncmanager.udl
Comment (lougeniaC64) [reply components/sync_manager/src/syncmanager.udl:57]: When I built android components locally against these changes while creating the `SyncManagerDeviceType` change, no `SyncReason` errors arose. It's worth noting that android defines it's own sync reasons and converts them to our sync reasons as needed (see [here](https://github.com/mozilla-mobile/firefox-android/blob/9de5ab8cc098674aad5190ce8b00ae6be65e9bd0/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/Types.kt#L44).

Comment (lougeniaC64) [reply components/sync_manager/src/syncmanager.udl:57]: Also please feel free to build a-c locally and check my math!

File: megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/RustSyncTelemetryPingTests.swift
Comment (lougeniaC64) [reply megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/RustSyncTelemetryPingTests.swift:14]: This absolutely does not cover all cases, and because we ultimately plan to remove this logic I didn't want to do too much here. I was most worried about the cases where a missing property is converted to default values (e.g. `EngineInfo.validation` or `OutgoingInfo`) but I can add cases for when errors are thrown, or if there are other cases people feel strongly about.

File: components/sync15/ios/RustSyncTelemetryPing.swift
Comment (lougeniaC64) [reply components/sync15/ios/RustSyncTelemetryPing.swift:60]: Yeah this and the `let` comment are fair points. I duplicated this code from `SyncTelemetryPing.swift` knowing that I needed most of it and adding the rest while I was there "just in case". But since we ultimately want to get rid of this serialization/deserialization logic and handle telemetry within a-s, I think stripping this file down to the bare essentials and treating it as a carrier of deserialized sync telemetry data makes a lot of sense.

File: components/sync15/ios/RustSyncTelemetryPing.swift
Comment (lougeniaC64) [reply components/sync15/ios/RustSyncTelemetryPing.swift:8]: See my comment below 😬

File: components/sync_manager/src/syncmanager.udl
Comment (tarikeshaq) [reply components/sync_manager/src/syncmanager.udl:57]: No makes sense! Thank you for checking 🙏

**Review by tarikeshaq — APPROVED** [review]
Sorry took so long to get back, LGTM! 🚀

**Review by mhammond — APPROVED** [review]
🚀

**Pull request merged**
